### DISCUSSION TITLE
Gaps between queryClient and useQuery API

### DISCUSSION THREAD

**Discussion by carpben** [TanStack/query#2549] [Ideas]
## The issue 
`queryClient.prefetch` doesn't accept and doesn't trigger `onSuccess` while `useQuery(fetchingObject)` does. 
This is an issue because: 

1. I think it is of value that the various fetching triggering consumers use a similar API. This can allow separating the logic for a specific fetch from the specific trigger. So if we are fetching for a post list  

```js
const postListQueryParams = {
  queryKey: "postList", 
  queryFn: ... 
  onSuccess: ...
}
```
we can then use the same object for the different triggers, such as `useQuery` and `queryClient.prefetch`.  
```jsx 
function Comp1(props){
  const queryResult = useQuery(postListQueryParams)
  ....
}

function Comp2(props){
  const queryResult = useQuery(postListQueryParams)
  ....
}

// Before Comp1 and Comp2 are in the UI 
queryClient.prefetch(postListQueryParams)
```

2. `onSuccess` could be very helpful when prefetching.

**Comment by TkDodo** [comment]
`onSuccess` is a parameter that only exists no an `observer` level, and since `prefetchQuery` (or even `fetchQuery`) don't create a subscription, there is no `onSuccess` callback. It's also briefly mentioned [in the docs here](https://react-query.tanstack.com/reference/QueryClient#queryclientfetchquery)

**Reply by carpben** [reply]
@TkDodo 
I was able to solve it using:
```js
queryClient.setQueryDefaults(queryKey ,{
   ... 
   onSuccess(data){
      ...
   }
})
```

When passed to `setQueryDefaults` `onSuccess` runs on prefetch. 
This solved my issue. However I find it inconsistent that `onSuccess` will run when passed to  `setQueryDefaults` but won't run when it is being passed directly to `prefetch`.

**Reply by TkDodo** [reply]
That’s indeed interesting. I’d need to take a closer look on why this is happening and what the expected behavior is.

**Reply by TkDodo** [reply]
is this just a type level issue? From what I can see, the options are gathered as usual from what you pass in and what the defaults are, so I think both ways _should_ work. But yeah, I can see that we don't have those on the typescript types 🤔

**Reply by carpben** [reply]
My experience was that the behaviour is different. When passing onSuccess directly to prefetch it didn't run, and when passing to to setQueryDefaults it did run. If this isn't your experience I'll be happy to try to create a CodeSandBox.

**Reply by TkDodo** [reply]
A sandbox would be great

**Reply by carpben** [reply]
I think I made a mistake. Perhaps because on one page `prefetch` ran alone, and on the other it ran parallel to `useQuery`.
